### ISSUE TITLE
"No such file or directory" after several writes when sass is watching

### ISSUE THREAD

**Issue by chearon** [sass/sass#394]
1.  Open a file and watch it
   
   ```
   $ sass --watch css/red.scss:css/red.css
   >>> Sass is watching for changes. Press Ctrl-C to stop.
   overwrite css/red.css
   WARNING: Listen has fallen back to polling, learn more at https://github.com/guard/listen#fallback.
   ```
   
   The warning happens on Ubuntu (probably in any Linux) ~~[because of a bug in Listen](https://github.com/guard/listen/issues/33)~~ (wasn't a bug, stupid mistake on my part, the latest version of Listen fixes it).  I got the crash described below regardless if that bug was patched or not.
2. Write to it several times
   `
   >>> Change detected to: css/red.scss
   overwrite css/red.css
   `
3.  After 4, 5, 20, or 30 times:
   `
   >>> Change detected to: css/red.scss
   >>> Deleted template detected: css/red.scss
    delete css/red.css
   /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/plugin/compiler.rb:330:in `read': No such file or directory - css/red.scss (Errno::ENOENT)
   from /var/lib/gems/1.8/gems/sass-3.1.16/vendor/listen/lib/listen/adapters/polling.rb:32:in `join'
   from /var/lib/gems/1.8/gems/sass-3.1.16/vendor/listen/lib/listen/adapters/polling.rb:32:in `start'
   from /var/lib/gems/1.8/gems/sass-3.1.16/vendor/listen/lib/listen/multi_listener.rb:44:in `start'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/plugin/listener.rb:30:in `start!'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/plugin/listener.rb:10:in `initialize'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/plugin/compiler.rb:257:in `new'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/plugin/compiler.rb:257:in `watch'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/plugin.rb:107:in `send'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/plugin.rb:107:in `method_missing'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/exec.rb:431:in `watch_or_update'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/exec.rb:302:in `process_result'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/exec.rb:41:in `parse'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/../lib/sass/../sass/exec.rb:21:in `parse!'
   from /var/lib/gems/1.8/gems/sass-3.1.16/bin/sass:8
   from /usr/local/bin/sass:19:in `load'
   from /usr/local/bin/sass:19
   $ 
   `
4.  Sass totally crashes.

**Comment by nex3** [comment]
And you are not in fact deleting the template?

**Comment by chearon** [comment]
I am not.  Just writing to it.

**Comment by nex3** [comment]
@Maher4Ever and @thibaudgg, do you have any insight? It looks like this may be a Listen issue.

**Comment by thibaudgg** [comment]
Mmm I think @Maher4Ever will be more in position to find this issue.

**Comment by Maher4Ever** [comment]
I tried to reproduce this issue on my machine with no luck. I even made a while loop which keeps updating the file at random intervals, but sass kept picking up changes.

@crh0872 Could you tell us more about your system (ubuntu version, ruby version)? Are you sure nothing else is watching the same file (say, Dropbox) or deleting it?

From the crash I was able to see that you are using ruby 1.8, so I even tried it on 1.8 and it still didn't crash.

Here are my system's info: (Ubuntu 12.04 32bit, ruby 1.8.7)

``` shell
U are in: /tmp/test 
» uname -a
Linux maher-laptop 3.2.0-24-generic-pae #37-Ubuntu SMP Wed Apr 25 10:47:59 UTC 2012 i686 i686 i386 GNU/Linux

U are in: /tmp/test 
» ruby -v
ruby 1.8.7 (2012-02-08 patchlevel 358) [i686-linux]
```

**Comment by chearon** [comment]
I think the key is Vim (I have the same kernel and Ruby).

I logged the events using `inotifywatch` and it looks like it's not actually Sass's fault, Vim just does a bunch of crazy things when you save a file.  Here's how Vim is saving the file `test.scss`:

```
test/ MOVED_FROM test.scss
test/ MOVED_TO test.scss~
test/ CREATE test.scss
test/ OPEN test.scss
test/ MODIFY test.scss
test/ CLOSE_WRITE,CLOSE test.scss
```

It moves the file to one ending in ~, then re-creates a new file with the original filename and copies the content back into it.  I think Sass sees the MOVE events as "file has been modified" and tries to do its thing then - sometimes by the time it is ready to open the file, Vim has moved it back, and less often, not.  Here's the output from the crash, which has some events sandwiched in between the MOVED test.scss and MODIFY test.scss.

```
test/ MOVED_FROM test.scss
test/ MOVED_TO test.scss~
test/ MODIFY .test.scss.swp
test/ MODIFY .test.scss.swp
test/ OPEN .test.scss.swp
test/ ACCESS .test.scss.swp
test/ ACCESS .test.scss.swp
test/ CLOSE_NOWRITE,CLOSE .test.scss.swp
test/ OPEN,ISDIR 
test/ CLOSE_NOWRITE,CLOSE,ISDIR 
test/ DELETE test.css
test/ CREATE test.scss
test/ OPEN test.scss
test/ MODIFY test.scss
test/ CLOSE_WRITE,CLOSE test.scss
test/ ATTRIB test.scss
test/ MODIFY .test.scss.swp
test/ DELETE test.scss~

```

**TL;DR**:  Setting `nowritebackup` in Vim causes it to behave normally: OPEN, MODIFY, CLOSE_WRITE.  I can't get sass to crash with that setting.  If this is indeed Vim's default behavior, I'm surprised nobody has come across it, but it might be useful to note this somewhere in the documentation or patch Listen for such behavior.

**Comment by thibaudgg** [comment]
FYI In Guard we have a wiki page about [inotify events for different editors](https://github.com/guard/guard/wiki/Analysis-of-inotify-events-for-different-editors).

**Comment by Maher4Ever** [comment]
The write backup mode in Vim is quite intreseting as it could to be a headache to get the right events out of it. Listen reports the MOVE events as add & remove events, so sass wouldn't get a modification event.

Here is what Listen reports for Vim:

```
modified: ["/tmp/test/.s.scss.swp"]
added:    ["/tmp/test/s.scss~"]
removed:  ["/tmp/test/s.scss"]

modified: []
added:    ["/tmp/test/s.scss"]
removed:  ["/tmp/test/s.scss~"]

modified: ["/tmp/test/.s.scss.swp"]
added:    []
removed:  []
```

Unfortunately, I tried to get sass to crash on me with Vim, but it didn't. I have a theory though why this crash occurs... In the output you posted it seems that sass crashes right after it detects the removal of `red.css`. In the inotify log you can see that at this point `red.scss` doesn't exist yet (I renamed the files in the log below to be consistent):

```
test/ DELETE red.css
test/ CREATE red.scss
```

If sass is doing something at that point with the scss file, that could be the source of the issue.

**Comment by chearon** [comment]
That's what I thought too.  The `DELETE red.css` looks like sass operating between Vim's move-scss-away and move-scss-back; but if it is, shouldn't we see something like `ACCESS red.scss`, `ATTRIB red.scss` or `OPEN red.scss`?  And I don't have any out of the ordinary Vim settings so I don't know why this isn't reproducible... it happens both on my 'buntu PC and laptop.

**Comment by shaldipur** [comment]
Hey guys, forgive me but I'm a bit of a noob with regards to this.  I saw that you guys were talking about this issue.  Here's what I got:

I'm getting the same issue.  Here it is:

Sachins-MacBook-Pro:app shaldipur$ sass --watch scss:css

> > > Sass is watching for changes. Press Ctrl-C to stop.
> > > WARNING: Listen has fallen back to polling, learn more at https://github.com/guard/listen#fallback.

I'm running sass (3.2.0.alpha.244) and ruby 1.8.7.  Installing the listen gem separately didn't help.  Installing ruby 1.9.3 through rvm didn't help either.  I saw somewhere that there is one that set polling to false.  Was anyone able to resolve this issue?  I have Dropbox also...would it be a good idea to temporarily un install it to see it it is causing the problems?

**Comment by thibaudgg** [comment]
@shaldipur Have you tried to update your Dropbox client?

**Comment by shaldipur** [comment]
I'm running the latest version...1.4.7

**Comment by Maher4Ever** [comment]
@shaldipur Do you have the `rb-fsevent` gem installed? sass uses pooling by default when there is no adapter installed on the system.
@nex3 Maybe you could add an info message telling people that sass doesn't come with Listen adapters and that they need to install it separately? Otherwise, people could be confused that something went wrong.

**Comment by shaldipur** [comment]
@Maher4Ever I don't have this one installed, but I recall having installed it when I installed the Listen gem separately.  I still got the same error.

**Comment by shaldipur** [comment]
Basically when I add rb-fsevent I get the following:

Sachins-MacBook-Pro:app shaldipur$ sass --watch scss:css

> > > Sass is watching for changes. Press Ctrl-C to stop.

So there isn't even a warning anymore.

**Comment by shaldipur** [comment]
I wonder if this is an issue with paths.   I did a test run with a file in the scss an css folders using absolute paths like so:

Sachins-MacBook-Pro:app shaldipur$ sass --watch /Users/shaldipur/StudentMentor.org/public_html/app/scss/account_areaofcategory.scss:/Users/shaldipur/StudentMentor.org/public_html/app/css/account_areaofcategory.css

> > > Sass is watching for changes. Press Ctrl-C to stop.
> > >       error /Users/shaldipur/StudentMentor.org/public_html/app/scss/sm_lib/_base.user.scss (Line 1: File to import not found or unreadable: sm_lib/base.
> > > Load path: /Users/shaldipur/StudentMentor.org/public_html/app)
> > > WARNING: Listen has fallen back to polling, learn more at https://github.com/guard/listen#fallback.

Then I ran a trace to see what was going on like so:

Running sass --watch --trace gets me the following:

Sachins-MacBook-Pro:app shaldipur$ sass --watch --trace
/Library/Ruby/Gems/1.8/gems/sass-3.2.0.alpha.244/bin/../lib/sass/../sass/exec.rb:366:in `watch_or_update': What files should I watch? Did you mean something like: (RuntimeError)
    sass --watch input.sass:output.css
    sass --watch input-dir:output-dir
    from /Library/Ruby/Gems/1.8/gems/sass-3.2.0.alpha.244/bin/../lib/sass/../sass/exec.rb:302:in`process_result'
    from /Library/Ruby/Gems/1.8/gems/sass-3.2.0.alpha.244/bin/../lib/sass/../sass/exec.rb:41:in `parse'
    from /Library/Ruby/Gems/1.8/gems/sass-3.2.0.alpha.244/bin/../lib/sass/../sass/exec.rb:21:in`parse!'
    from /Library/Ruby/Gems/1.8/gems/sass-3.2.0.alpha.244/bin/sass:8
    from /usr/bin/sass:23:in `load'
    from /usr/bin/sass:23

So, I'm more or less lost. Seems like I'm  missing something but I can't figure out what it could be.

**Comment by thibaudgg** [comment]
@shaldipur could you try the solution given in issue #362: https://github.com/nex3/sass/issues/362#issuecomment-6112881

**Comment by shaldipur** [comment]
I think its running now. Thanks!

**Comment by paperscissors** [comment]
This happens to me, but I think it is because I am editing live on the server via Transmit. I think potentially Transmit deletes the existing file and replaces it with a new upload, so this error doesn't surprise me. The best I could think to rectify this would be to increase the time buffer for when SASS will tank if the file is deleted, so instead of it stopping right away if the file is missing, maybe note the file is missing, and if it's not back after five seconds, then stop?

**Comment by chriseppstein** [comment]
Closing: duplicate of #362
